pennylane/templates/subroutines/grover.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the Grover Operation template.
+"""
+import pennylane as qml
+from pennylane.operation import AnyWires, Operation
+from pennylane.ops import Hadamard, PauliZ, MultiControlledX
+
+
+class GroverOperator(Operation):
+    r"""Performs the Grover Diffusion Operator.
+
+    .. math::
+
+        G = 2 |s \rangle \langle s | - I
+        = H^{\bigotimes n} \left( 2 |0\rangle \langle 0| - I \right) H^{\bigotimes n}
+
+
+    For this template, the operator is implemented with a layer of Hadamards, an
+    effective multi-controlled Z gate, and another layer of Hadamards.
```

Comment:
Recommend to align the picture with the description; the diagram skips a couple of simplification steps that happen when you apply the identity `HZH = X`. One option is to have a new diagram just showing the circuit as described here, but then that wouldn't match the implementation. So maybe for clarity both versions can be shown, with an equals sign in the middle and brief mention of how you get one from the other.

In either case, I think it'd be good to mention that the open-circle notation means to control on 0 (or alternatively, have a layer of X gates).